### PR TITLE
Travis CI: Use flake8 to find Python syntax errors

### PR DESCRIPTION
**Pull request by cclauss**
In Travis CI, add Python 2 and Python 3 linting steps that run [flake8](http://flake8.pycqa.org) to find syntax errors and undefined names.

__E901,E999,F821,F822,F823__ are the "_showstopper_" flake8 issues that can halt the runtime with a SyntaxError, NameError, etc. Most other flake8 issues are merely "style violations" -- useful for readability but they do not effect runtime safety.
* F821: undefined name `name`
* F822: undefined name `name` in `__all__`
* F823: local variable name referenced before assignment
* E901: SyntaxError or IndentationError
* E999: SyntaxError -- failed to compile a file into an Abstract Syntax Tree

### CODE REVIEW

**Comment by cclauss** [comment]
```
Python 2.7:
32    E999 SyntaxError: invalid syntax
25    F821 undefined name 'actual'
57

Python 3.6:
24    E999 SyntaxError: invalid syntax
27    F821 undefined name 'actual'
51
```

**Comment by arnavb** [comment]
@cclauss We now use black for autoformatting so flake8 might not be the best choice; I would close this. However, pylint might not be a bad idea to add.

**Comment by cclauss** [comment]
syntax errors and undefined names have nothing to do with formatting and black does nothing to fix them.

**Comment by arnavb** [comment]
I see what you mean. However, since we're only focusing on errors related to code smells and various syntax errors, wouldn't [pylint](https://github.com/PyCQA/pylint) be a better fit? From my experience, pylint catches more code smells and other errors than flake8. Flake8, I believe, is more focused on style and while this can be disabled, it doesn't seem like the right tool for the job to me (the [official docs](flake8.pycqa.org/en/latest/#) advertise it as "Your Tool For Style Guide Enforcement").

File: .travis.yml [hunk 1]
```diff
@@ -54,6 +54,28 @@ matrix:
       env:
         - MATRIX_EVAL="CXX=g++-7"
 
+    # Lint Python 2 with flake8
+    - os: linux
```
Comment (arnavb) [inline .travis.yml]: I don't think we need to keep Python 2 support. Yes, some of the algorithms contain Python 2 code, but they _should_ have been Python 3 in the first place. I would remove this matrix entry.

File: .travis.yml [hunk 2]
```diff
@@ -54,6 +54,28 @@ matrix:
...
+      python: 2.7
+      before_install: pip install flake8
+      script:
+        # stop the build if there are Python 2 syntax errors or undefined names
+        - flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics
+        # exit-zero treats all errors as warnings.  The GitHub editor is 127 chars wide
+        - flake8 . --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics
+
+    # Lint Python 3 with flake8
+    - os: linux
```
Comment (arnavb) [inline .travis.yml]: Could you give this job a name (such as `Python linting`)? If you do so, please add it to the other Python job as well (the autoformat one).

**Comment by arnavb** [comment]
@cclauss Will you update?

**Pull request closed without merging**